File: SuperGol/src/main/java/ar/edu/unq/model/Team.java
package ar.edu.unq.model;

import ar.edu.unq.model.position.Position;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;


public class Team {

    private String name;
    private User owner;
    private Set<Player> players;
    private String logo;
    private Player captain;


    public Team(String aName, User aUser){
        name = aName;
        owner = aUser;
        players = new HashSet<>();
    }

    public Team(String aName, User aUser, String teamLogo) {
        this(aName, aUser);
        logo = teamLogo;
    }

    public void addPlayer(Player aPlayer) throws Exception {
        Position position = aPlayer.getPosition();
        if(getPlayersWithPosition(position).size() == (position.getMaxNumberOfPlayersInATeam()))
            throw new Exception("You should remove a " + position + " before adding another one");

        players.add(aPlayer);
    }

    public void setCaptain(Player aPlayer) throws Exception {
        validatePlayerIsInTheTeam(aPlayer);

        captain = aPlayer;
    }

    //Getters - Setters
    public String getName() {
        return name;
    }

    public void setName(String aName) {
        name = aName;
    }

    public String getLogo() {
        return logo;
    }

    public void setLogo(String aLogo) {
        logo = aLogo;
    }

    public Set<Player> getPlayers() {
        return players;
    }

    public User getOwner() {
        return owner;
    }

    //Private
    private List<Player> getPlayersWithPosition(Position aPosition) {
        return players.stream()
                .filter(player -> player.getPosition().equals(aPosition))
                .collect(Collectors.toList());
    }

    private void validatePlayerIsInTheTeam(Player aPlayer) throws Exception {
        if(!players.contains(aPlayer))
            throw new Exception(aPlayer.getName() + "doesn't belong to this team");
    }

    public Integer pointsMadeIn(DateMatch aDateMatch) {
        return players.stream()
                .map(player -> player.pointsMadeIn(aDateMatch))
                .reduce(0, (a, b) -> a + b);
    }

    public Player getCaptain() {
        return captain;
    }
}

File: super_goal/app/models/team.rb
class Team < ActiveRecord::Base

  attr_reader :captain
  has_many :players
  has_one :captain, class_name: 'Player'

  def add_player(a_player)
    validate_there_is_room_for a_player

    players << a_player
  end

  def has_player?(a_player)
    players.include? a_player
  end

  def assign_captain(a_player)
    validate_player_is_in_the_team a_player

    @captain = a_player
  end

  def points_made_in(a_date_match)
    players.collect { |player| player.points_made_in a_date_match }
            .sum
  end

  protected

  def validate_player_is_in_the_team(a_player)
    raise "Couldn't assign #{a_player.name} as captain since he doesn't belong to the team" unless players.include? a_player
  end

  def validate_there_is_room_for(a_player)
    position = a_player.position
    if players_with_position(position).size.equal?(position.max_number_of_players_per_team)
      raise "You must remove a #{position.name} before adding another one"
    end
  end

  def players_with_position(a_position)
    players.select { |player| player.has_position? a_position }
  end
end